ls1intum/Artemis
# Feature: Provide multiple templates for a programming language

### Checklist

### Motivation and Context
Currently, there can only be one template per programming exercise. For Java, often two exercise types are used: Eclipse projects and Maven projects. Instructors currently have to manually adjust the exercise if they want to use Maven exercises.
See #2062.

### Description
This PR adds an attribute `projectType` to the Programming Exercises, defaulting to `null`. Each programming language has a base template (consisting of the already present folder `exercise`, `solution`, `tests`, which can be extended by project-specific files (e.g. `pom.xml` or `.project` files). The selection is not shown if there is only one template for this programming language.

The Project Type can not be changed after creating the exercise as the repositories are setup with the selected option during exercise creation.

### Steps for Testing
1. Create a new programming exercise
2. See the new selection for the project type when select Java as programming language.
3. See the difference in the loaded README when selecting Maven instead of Eclipse
4. Create one project using Maven and one using Eclipse.
- The eclipse project should be identical to the previous template.
- The maven project should be identical except that 1) a pom.xml file is generated for student instead of `.classpath` and `.project` file. 2) The `Client` class uses a method from Apache RandomUtils.
5. Import both exercises into eclipse (one as eclipse project and one as maven project), you should be able to run the `main` method in the `Client` class.
6. Create another non-Java Programming exercise: The project type selection should not be present but you should still be able to create the exercise using the default template.

### Test Coverage

* ExerciseService.java: 85%
* programming-exercise.component.ts 95%

### Screenshots
![Peek 2020-10-26 18-03](https://user-images.githubusercontent.com/5084100/97203937-c49bdb00-17b5-11eb-8fb1-d8b3887f8c23.gif)
